tests/test_bootstrap.py

```diff
@@ -0,0 +1,551 @@
+
+    def test_get_bootstraps_from_recipes(self):
+        """A test which will initialize a bootstrap and will check if the
+        method :meth:`~pythonforandroid.bootstrap.Bootstrap.
+        get_bootstraps_from_recipes` returns the expected values
+        """
+        recipes_service = {"pyjnius", "python3"}
+        bs = Bootstrap().get_bootstrap_from_recipes(recipes_service, self.ctx)
+        self.setUp_distribution_with_bootstrap(bs)
+
+        self.assertEqual(bs.name, "service_only")
+
+        # test wrong recipes
+        wrong_recipes = {"python2", "python3", "pyjnius"}
+        bs = Bootstrap().get_bootstrap_from_recipes(wrong_recipes, self.ctx)
+        self.assertIsNone(bs)
+
+    @mock.patch("pythonforandroid.bootstrap.ensure_dir")
+    def test_prepare_dist_dir(self, mock_ensure_dir):
+        """A test which will initialize a bootstrap and will check if the
+        method :meth:`~pythonforandroid.bootstrap.Bootstrap.prepare_dist_dir`
+        successfully calls once the method `endure_dir`
+        """
+        # prepare bootstrap and distribution
+        bs = Bootstrap().get_bootstrap("sdl2", self.ctx)
+        self.setUp_distribution_with_bootstrap(bs)
+
+        bs.prepare_dist_dir("fake_name")
+        mock_ensure_dir.assert_called_once_with(bs.dist_dir)
+
+    @mock.patch("pythonforandroid.bootstrap.open", create=True)
+    @mock.patch("pythonforandroid.util.chdir")
+    @mock.patch("pythonforandroid.bootstrap.shutil.copy")
+    @mock.patch("pythonforandroid.bootstrap.os.makedirs")
+    @mock.patch("pythonforandroid.bootstrap.sh.ln")
+    @mock.patch("pythonforandroid.bootstrap.listdir")
+    @mock.patch("pythonforandroid.bootstrap.sh.mkdir")
+    @mock.patch("pythonforandroid.bootstrap.sh.rm")
+    @mock.patch("pythonforandroid.bootstrap.isdir")
+    def test_bootstrap_prepare_build_dir(
+        self,
+        mock_isdir,
+        mock_sh_rm,
+        mock_sh_mkdir,
+        mock_listdir,
+        mock_sh_ln,
+        mock_os_makedirs,
+        mock_shutil_copy,
+        mock_chdir,
+        mock_open,
+    ):
+        """A test which will initialize a bootstrap and will check if the
+        method :meth:`~pythonforandroid.bootstrap.Bootstrap.prepare_build_dir`
+        successfully calls the methods that we need to prepare a build dir.
+        """
+        mock_isdir.return_value = True
+        mock_sh_mkdir.return_value = "sh.mkdir successfully mocked"
+        mock_sh_rm.return_value = "sh.rm successfully mocked"
+        mock_listdir.return_value = [
+            "jnius",
+            "kivy",
+            "Kivy-1.11.0.dev0-py3.7.egg-info",
+            "pyjnius-1.2.1.dev0-py3.7.egg",
+        ]
+        mock_sh_ln.return_value = True
+        mock_os_makedirs.return_value = True
+        mock_shutil_copy.return_value = True
+        mock_chdir.return_value = True
+        mock_open.side_effect = [
+            mock.mock_open(read_data="target=android-21").return_value
+        ]
```

Comment:
This is specifically confusing me :thinking:

Reply:
Aaa...yea...we need to do this because we test the `open` inside a Context Manager and the default Mock object lacks of  the attributes `__enter__` and `__exit__` wich will lead into an error, adding into the final `.return_value` solves the problem...
maybe do you have a better approach?

Reply:
Which error? Because it's a magic mock to me and so is your return_value here.
Personally I would drop all the `mock_open.side_effect....` part. The tests should still pass without it and I can't see why it's needed :thinking:

Reply:
@AndreMiras,You are right, I was trying to reinvent the wheel, we don't need that line **thanks!!** :smile: